presto-hive/src/main/java/com/facebook/presto/hive/s3/PrestoS3FileSystem.java

```diff
@@ -297,20 +297,21 @@ public LocatedFileStatus next()
     public FileStatus getFileStatus(Path path)
             throws IOException
     {
+        String key = keyFromPath(path);
+        ObjectMetadata metadata = getS3ObjectMetadata(key);
+
         if (path.getName().isEmpty()) {
             // the bucket root requires special handling
-            if (getS3ObjectMetadata(path) != null) {
+            if (metadata != null) {
                 return new FileStatus(0, true, 1, 0, 0, qualifiedPath(path));
             }
             throw new FileNotFoundException("File does not exist: " + path);
         }
 
-        ObjectMetadata metadata = getS3ObjectMetadata(path);
-
         if (metadata == null) {
             // check if this path is a directory
             Iterator<LocatedFileStatus> iterator = listPrefix(path);
-            if (iterator.hasNext()) {
+            if (iterator.hasNext() || getS3ObjectMetadata(key + PATH_SEPARATOR) != null) {
```

Comment:
I may be missing something here, but, say `path` is `/a`, `listPrefix` will list the objects with prefix `/a/`  (`listPrefix` appends `PATH_SEPARATOR` to the given `key`). If the returned iterator has no elements (so no objects exist with a key that starts with the prefix `/a/`), then how would `getS3ObjectMetadata("/a/")` return a non-null value?

Reply:
The problem is that `listPrefix` method filters out directories. I agree, that this implementation is not intuitive. Reverted it and splitted `listPrefix` method into two parts instead.